tensorflow/core/kernels/sparse_concat_op.cc

```diff
@@ -99,54 +107,166 @@ class SparseConcatOp : public OpKernel {
                   " at position ", i));
         }
       }
+      out_shape_vec[concat_dim] += shapes[i].vec<int64>()(concat_dim);
+      concat_dim_offsets[i] =
+          concat_dim_offsets[i - 1] + shapes[i - 1].vec<int64>()(concat_dim);
+      val_offsets[i] = val_offsets[i - 1] + inds[i - 1].dim_size(0);
+      num_entries += inds[i].dim_size(0);  // Update number of entries
+    }
+
+    // In this implementation, the concat process has 3 steps:
+    // 1. Copy all tensor indices and values to an output with the input order,
+    // and re-calculate the indices after concatation.
+    // 2. Perform a sort on indices to get the final position of each element.
+    // 3. Copy the indices and values to the final output tensor according to
+    // the sorting result.
+    Tensor output_ix(DT_INT64, TensorShape({num_entries, input_rank}));
+    Tensor output_vals(DataTypeToEnum<T>::v(), TensorShape({num_entries}));
+    Tensor output_shape(DT_INT64, TensorShape({input_rank}));
+    auto output_shape_t = output_shape.vec<int64>();
+    std::copy_n(out_shape_vec.begin(), input_rank, output_shape_t.data());
+
+    auto worker_threads = context->device()->tensorflow_cpu_worker_threads();
+    auto workers = worker_threads->workers;
```

Comment:
could these be references?

Reply:
Well, fixed with explicit types with reference. Thx.